shard-operator/src/main/java/com/redhat/service/bridge/shard/operator/monitoring/ServiceMonitorServiceImpl.java

```diff
@@ -0,0 +1,69 @@
+package com.redhat.service.bridge.shard.operator.monitoring;
+
+import java.util.Optional;
+
+import javax.enterprise.context.ApplicationScoped;
+import javax.inject.Inject;
+
+import org.slf4j.Logger;
+import org.slf4j.LoggerFactory;
+
+import com.redhat.service.bridge.shard.operator.providers.TemplateProvider;
+import com.redhat.service.bridge.shard.operator.utils.LabelsBuilder;
+
+import io.fabric8.kubernetes.api.model.LabelSelector;
+import io.fabric8.kubernetes.api.model.Service;
+import io.fabric8.kubernetes.client.CustomResource;
+import io.fabric8.kubernetes.client.KubernetesClient;
+import io.fabric8.openshift.api.model.monitoring.v1.ServiceMonitor;
+
+@ApplicationScoped
+public class ServiceMonitorServiceImpl implements ServiceMonitorService {
+
+    private static final Logger LOGGER = LoggerFactory.getLogger(ServiceMonitorServiceImpl.class);
+
+    @Inject
+    KubernetesClient kubernetesClient;
+
+    @Inject
+    TemplateProvider templateProvider;
+
+    @Override
+    public Optional<ServiceMonitor> fetchOrCreateServiceMonitor(final CustomResource resource, final Service service) {
+        if (!ServiceMonitorClient.isServiceMonitorAvailable(kubernetesClient)) {
```

Comment:
@robpblake, I think we can deploy everything and not fail, but add in the conditions something like:

| Type | Status | Reason | Message | 
| ----- | ------ | -------- | ---------|
| Ready | False | PrometheusUnavailable | ServiceMonitor CRD not available in the cluster | 

This way I believe the SREs will have valuable feedback and at the same time, the service is there, serving. :)

But misconfiguration won't be noticed by our side unless we poke the Prometheus API. See: https://prometheus-operator.dev/docs/operator/troubleshooting/#has-my-servicemonitor-been-picked-up-by-prometheus

Reply:
@ricardozanini Thinking about it I actually think this is a bigger discussion around dependency management of the Shard Operator. I think the Shard Operator should as part of its boot/install sequence, check that it has all dependencies required and fail to start if they are not available.

My assumption being that we will delegate the dependency checking and management to OLM so we can list Prometheus as a GVK dependency that the Fleet Shard Operator has.

My concern about not failing fast on lack of runtime dependencies is the recovery procedure for the SRE teams. If we have ~hundreds of Ingresses in a particular OSD, then it's going to be pretty cumbersome...

My suggestion is that we make the assumption that Prometheus is available in the cluster and we then open up another ticket to make our operator installable via OLM.

Your thoughts?

Reply:
> My assumption being that we will delegate the dependency checking and management to OLM so we can list Prometheus as a GVK dependency that the Fleet Shard Operator has.

That's perfectly fine. However, we can't forget the local environment where I don't want to install Prometheus in order to run the operator. Here I can see two simple options: (a) we add the warning in the conditions, or; (b) we skip the `ServiceMonitor` installation based on application properties. Here I believe it's worth mapping this file to a configMap so SREs could control the operator behavior.

> My concern about not failing fast on lack of runtime dependencies is the recovery procedure for the SRE teams. If we have ~hundreds of Ingresses in a particular OSD, then it's going to be pretty cumbersome...

We do file fast. The `Ready: false` condition is there for this reason. With observability configured in the cluster, they can perfectly set alerts based on this condition and react. So, I don't think we should fail the deployment entirely. But, this is not my drum to beat. So, let's take this use case to them and reassess it. 

> My suggestion is that we make the assumption that Prometheus is available in the cluster so that the Shard does not have to optionally create the ServiceMonitor and we then create another ticket in our backlog to make our Fleet Shard operator installable via OLM.

+1 to add Prometheus dependency on OLM installations since this won't interfere with local deployments.

Reply:
Thanks @ricardozanini . Some good points there.

> we can't forget the local environment where I don't want to install Prometheus in order to run the operator. Here I can see two simple options: (a) we add the warning in the conditions,

Good point. Let's go with the Conditions for now then seeing as you've already done the work.

> But, this is not my drum to beat. So, let's take this use case to them and reassess it.

+1  I'll create us a JIRA to track this discussion with SRE teams.

> +1 to add Prometheus dependency on OLM installations since this won't interfere with local deployments.

+1. I'll create us a ticket to consider OLM packaging of the Fleet Shard Operator. Not a _huge_ priority, but we'll need it at some point.

cheers!